kythe/java/com/google/devtools/kythe/util/schema/Nodes.java

```diff
@@ -120,9 +129,67 @@ public static Node normalizeNode(Node n) {
                 b.addFact(f);
               }
             });
+  }
+
+  /**
+   * Returns a {@link Node} from the given set of {@link Entry} protos. All {@link Entry}s must
+   * share the same source {@link VName}.
+   */
+  public static Node fromEntries(Iterable<Entry> entries) {
+    List<Fact> facts = new ArrayList<>();
+    List<Edge> edges = new ArrayList<>();
+    for (Entry e : entries) {
+      if (e.hasFact()) {
+        facts.add(e.getFact());
+      } else {
+        edges.add(e.getEdge());
+      }
+    }
+    final Node.Builder b = Node.newBuilder();
+    if (!facts.isEmpty() && facts.get(0).hasSource()) {
+      b.setSource(facts.get(0).getSource());
+    } else if (!edges.isEmpty() && edges.get(0).hasSource()) {
+      b.setSource(edges.get(0).getSource());
+    }
+    addEntries(b, facts, edges);
     return b.build();
   }
 
+  /** Returns the composite {@link Entry} protos of the given {@link Node}. */
+  public static Iterable<Entry> toEntries(Node node) {
+    String nodeKind = getNodeKind(node);
+    String subkind = getSubkind(node);
+    Stream<Fact> facts =
+        Streams.concat(
+            nodeKind.isEmpty()
+                ? Stream.of()
+                : Stream.of(
+                    Fact.newBuilder()
+                        .setKytheName(FactName.NODE_KIND)
+                        .setValue(ByteString.copyFromUtf8(nodeKind))
+                        .build()),
+            subkind.isEmpty()
+                ? Stream.of()
+                : Stream.of(
+                    Fact.newBuilder()
+                        .setKytheName(FactName.SUBKIND)
+                        .setValue(ByteString.copyFromUtf8(subkind))
+                        .build()),
+            node.getFactList().stream());
+    Stream<Edge> edges = node.getEdgeList().stream();
+    if (node.hasSource()) {
+      facts = facts.map(f -> f.toBuilder().setSource(node.getSource()).build());
+      edges = edges.map(e -> e.toBuilder().setSource(node.getSource()).build());
+    } else {
+      facts = facts.map(f -> f.toBuilder().clearSource().build());
+      edges = edges.map(e -> e.toBuilder().clearSource().build());
+    }
+    return Streams.concat(
+            facts.map(f -> Entry.newBuilder().setFact(f).build()),
+            edges.map(e -> Entry.newBuilder().setEdge(e).build()))
+        ::iterator;
```

Comment:
Streams are single-pass, Iterables are multi-pass.  This should probably be:
```
return () -> {
  ...
  return Streams.concat(...).iterator();
};
```

(There was an internal cleanup along this lines, see https://github.com/kythe/kythe/pull/4639 for the application of that)

Reply:
(or return a Stream rather than an iterable and let callers iterate it or not).